src/pkcs15init/pkcs15-openpgp.c

```diff
@@ -185,46 +202,167 @@ static int openpgp_generate_key(sc_profile_t *profile, sc_pkcs15_card_t *p15card
-	pubkey->u.rsa.exponent.data = calloc(key_info.rsa.exponent_len>>3, 1); /* 1/8 */
+	pubkey->u.rsa.exponent.len = key_info.u.rsa.exponent_len;
+	pubkey->u.rsa.exponent.data = calloc(BYTES4BITS(key_info.u.rsa.exponent_len), 1);
 	if (pubkey->u.rsa.exponent.data == NULL)
-		goto out;
-	memcpy(pubkey->u.rsa.exponent.data, key_info.rsa.exponent, key_info.rsa.exponent_len>>3); /* 1/8 */
-
-out:
-	if (key_info.rsa.modulus)
-		free(key_info.rsa.modulus);
-	if (key_info.rsa.exponent)
-		free(key_info.rsa.exponent);
+		goto err;
+	memcpy(pubkey->u.rsa.exponent.data, key_info.u.rsa.exponent, BYTES4BITS(key_info.u.rsa.exponent_len));
+
+err:
+	if (key_info.u.rsa.modulus)
+		free(key_info.u.rsa.modulus);
+	if (key_info.u.rsa.exponent)
+		free(key_info.u.rsa.exponent);
+	LOG_FUNC_RETURN(ctx, r);
+}
+
+
+/**
+ * Generates a new ECC key pair using an existing key file.
+ * @param  card     IN sc_card_t object to use
+ * @param  obj      IN sc_pkcs15_object_t object with pkcs15 information
+ * @param  pukkey   OUT the newly created public key
+ * @return SC_SUCCESS on success and an error code otherwise
+ **/
+static int openpgp_generate_key_ec(sc_card_t *card, sc_pkcs15_object_t *obj,
+									sc_pkcs15_pubkey_t *pubkey)
+{
+	sc_context_t *ctx = card->ctx;
+	sc_cardctl_openpgp_keygen_info_t key_info;
+	sc_pkcs15_prkey_info_t *required = (sc_pkcs15_prkey_info_t *)obj->data;
+	sc_pkcs15_id_t *kid = &(required->id);
+	const struct sc_ec_parameters *info_ec =
+	    (struct sc_ec_parameters *) required->params.data;
+	unsigned int i;
+	int r;
+
+	LOG_FUNC_CALLED(ctx);
+	memset(&key_info, 0, sizeof(key_info));
+
+	sc_log(ctx, "Key ID to be generated: %s", sc_dump_hex(kid->value, kid->len));
+
+	/* Accept KeyID = 45, which is default value set by pkcs15init */
+	if (kid->len == 1 && kid->value[0] == 0x45) {
+		/* Default key is authentication key. We choose this because the common use
+		 * is to generate from PKCS#11 (Firefox/Thunderbird) */
+		sc_log(ctx, "Authentication key is to be generated.");
+		key_info.key_id = 3;
+	}
+	if (!key_info.key_id && (kid->len > 1 || kid->value[0] > 3)) {
+		sc_log(ctx, "Key ID must be 1, 2 or 3!");
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+
+	if (!key_info.key_id)
+		key_info.key_id = kid->value[0];
+
+
+	/* set algorithm id based on key reference */
+	key_info.algorithm = (key_info.key_id == SC_OPENPGP_KEY_ENCR)
+			   ? SC_OPENPGP_KEYALGO_ECDH /* ECDH for slot 2 only */
+			   : SC_OPENPGP_KEYALGO_ECDSA; /* ECDSA for slot 1 and 3 */
+
+	/* extract oid the way we need to import it to OpenPGP Card */
+	if (info_ec->der.len > 2)
+		key_info.u.ec.oid_len = info_ec->der.value[1];
+	else
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	for (i=0; (i < key_info.u.ec.oid_len) && (i+2 < info_ec->der.len); i++){
+		key_info.u.ec.oid.value[i] = info_ec->der.value[i+2];
+	}
+	key_info.u.ec.oid.value[key_info.u.ec.oid_len] = -1;
+
+	/* Prepare buffer */
+	key_info.u.ec.ecpoint_len = (required->field_length>>2) + 1;
```

Comment:
Honestly, you really got me here... I needed to think about this for a *really* long time and realized, that I probably just understood something wrong when I did this. Funnily I reversed it in card-openpgp.c afterwards anyway as you can see in the commit.

I looked through the usage of ecpoint_len very carefully and now I am quite confident that the issue is fixed.

Thanks for looking at the code thorougly.

Reply:
Does this mean that this code fragment had an undetected bug? Did you verify the correctness with your new change?

Reply:
In part, I would say it did.

The problem is that I am not able to fully test the *correctness* of the results as e.g. storing an ec key fails on behalf of OpenSC. Key generation on card worked and works fine though. This is due to the fact that I first divided the length by four and then tricked myself by multiplying 4 again (as seen in the last commit). I am sorry for this mistake.

Reply:
I see. Then, however, I'm asking myself if we need this code at all if nobody ever looks into what you've encoded on the card. Could this be verified with gpg?

Reply:
When I started to implement ECC key support I assumed that importing ECC keys is working in OpenSC. Thus I implemented both, key generation and import alike with all needed subfunction. We could disable the key import but I felt like it would not hurt to have the already implemented code until the import is working for OpenSC.

In general I - of course - excpect the code to work. I am not so sure about the fingerprint function and there might be some other issues though.

I can not compare/verify with gpg as long as I can not import keys. Even then it is a tricky task as there is no ready-to-go way to move a S/MIME based key to OpenPGP standard. (Currently, I am looking for a way to get this a bit more fluid in the long run.)

Reply:
If you plan to add the missing features in the future, then we can already merge the prerequisite now.

Reply:
That would be great. I can not promise that I'll be able to do this anytime soon though, I am afraid.